### ISSUE TITLE
performance: Stat files in parallel

### ISSUE THREAD

**Issue by nh2** [rclone/rclone#7153]
#### What is your current rclone version (output from `rclone version`)?

`1.62.2`

#### What problem are you are trying to solve?

Operations such as `rclone copy` that need to determine what files are on the disk are slower than they need be because file statting (e.g. the `newfstatat()` syscall on Linux) is done sequentially.

Many storage devices today benefit from parallel filesystem metadata access:

* Multi-spinning disk storage servers and NASs have multiple disks that can seek in parallel.
* Current SSDs perform fastest when queried in parallel.
* Network-backed storage, as is common on cloud providers for block storage backing VMs (e.g. Amazon EBS), or networked file systems (NFS, Ceph), have network latency that slows down serial access.

For all of these cases, access latency slows down `stat()` calls when done serially.

This latency can be hidden by doing many of them in parallel.

Concrete example:

* I have a server with 10 spinning disks, 50M files.
* When `rclone copy` gathers file sizes and modification times, does 50M `newfstatat()` syscalls, serially from 1 thread (shown by `strace`).
* It does 250 syscalls per second, because the average seek time on a single spinning disk is 4ms.
* This means that it takes `50e6 / 250 / 3600` = **55 hours** until `rclone copy` starts any copying.
* If rclone used all 10 disks in parallel, it would be 10x faster.

When dealing with an SSD-backed cloud storage device, the difference between serial and concurrent access is even more extreme, and speedups closer to 100x or 1000x can be achieved.

#### How do you think rclone should be changed to solve that?

rclone should add a flag `--stat-threads int` (similar to `--checkers int`) with which the user can control how many stats to do in parallel.

A similar feature was implemented in the `bupstash` backup tool ([PR](https://github.com/andrewchambers/bupstash/commit/5be2c885b5c4d6904291a7d2d6e406421b90d38f#diff-b8ebb30fb413ab1236cbf2f2a2a5690a7a1992759293208b65962779b39e5eb7R201), which achieved great speedups.

**Comment by ncw** [comment]
I think @darthShadow some ideas about this.

I keep hoping to find an interface like that on Windows which reads the directory listings and the stats at the same time.

Given that doesn't exist, parallel stats sounds like a great plan and not too difficult to implement.

**Comment by nh2** [comment]
> I keep hoping to find an interface like that on Windows which reads the directory listings and the stats at the same time.
>
> Given that doesn't exist

Indeed, Linux does not have such a thing. The `getdents()` and `stat()` syscall families strictly separate getting the list of file names from getting per-file metadata, also due to the way that most Linux file systems store this info internally.

**Comment by darthShadow** [comment]
I think it would be better to keep the discussion in a single issue: https://github.com/rclone/rclone/issues/6632

My next aim is to implement something like kopia's parallel listing ([kopia/kopia@8b0296c/fs/localfs/local_fs.go#L129](https://github.com/kopia/kopia/blob/8b0296cdf2c37d4f0f3de408eb653b80b56a23a5/fs/localfs/local_fs.go#L129)) which should improve performance significantly.

**Comment by nh2** [comment]
> I think it would be better to keep the discussion in a single issue: #6632

@darthShadow Makes sense, but I think you should rename that issue's title in that case.

When searching for existing issues, I scrolled by it, but didn't consider it because the description only mentions VFS, not local normal statting (which is indeed discussed in that issue).
